src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt

```diff
@@ -56,9 +56,15 @@ open class GossipRouter @JvmOverloads constructor(
     val acceptRequestsWhitelistMaxMessages = 128
     val acceptRequestsWhitelistDuration = 1.seconds
 
-    val score by lazy { GossipScore(scoreParams, executor, curTimeMillis) }
+    open val score: GossipScore by lazy {
+        DefaultGossipScore(scoreParams, executor, curTimeMillis).also {
+            eventBroadcaster.listeners += it
+        }
+    }
```

Comment:
It works just because the `score` in lazy initialized: the code inside `lazy` block is executed on the first `score` usage
TBH I was trying to invent 2-in-1 'builder+class' pattern with Kotlin but it's far from what it's intended to be and tends to spaghetti-code pattern

Reply:
Was trying to refactor that quickly with the classic Builder pattern but seemed to open a can of worms so it's definitely worth a separate PR

Reply:
ah right - I hadn't spotted the `lazy` there at all, mostly because it's a kotlin feature so I'm just not used to it. I'm fine leaving it as-is but sounds like you're not happy with the pattern anyway so go with what you think.